Enable image optimization setting by default

SiobhyB opened the pull request:
## Description

This PR changes the default setting for `Optimize Images` (found via _Me_ → _App Settings_) from off → on. We will also inform users of this new default via a popup that will display when uploading media.

For now, the default Maximum Image Size is _3000px_ and the Image Quality is _Medium_. These defaults are up for discussion and may change.

You can find the iOS equivalent at https://github.com/wordpress-mobile/WordPress-iOS/pull/21981.

| New pop-up  | Default settings |
| ------------- | ------------- |
| <img src="https://github.com/wordpress-mobile/WordPress-Android/assets/2998162/787e4e56-9043-431c-ade2-eccd47681e35" width="100%"> | <img src="https://github.com/wordpress-mobile/WordPress-Android/assets/2998162/308712a2-4bd7-4b73-94b6-c817a2af3278" width="100%"> |

> **Note**
> Iterations that have been identified for future PRs include enabling the popup when a photo is taken from the Story block and additional automated tests.

## To test

<details>
 <summary>Verify default setting displays as expected ⤵️</summary>

### Default setting should be changed if it has never been set before

* Perform a fresh installation of the app.
* Navigate to _Me → App Settings_. 
* Observe that _Optimize Images_ setting is enabled by default.
* Observe that _Maximum Image Size_ and _Image Quality_ settings have the default values as shared in the PR's description.

### Default setting should not be changed if it has been set previously

#### Set original as maximum image size
* Install an older version.
* Navigate to _Me → App Settings_. 
* Move the _Maximum Image Size_ setting to a different value than the default.
* Move the _Maximum Image Size_ setting back to the original value.
* Install the installable build from this PR.
* Navigate to _Me → App Settings_. 
* Observe that the _Optimize Images_ setting is enabled and that _Maximum Image Size_ setting remains in the previously set original value.

#### Set maximum image size
* Install an older version.
* Navigate to _Me → App Settings_. 
* Move the _Maximum Image Size_ setting to a different value than the original.
* Install the installable build from this PR.
* Navigate to _Me → App Settings_. 
* Observe that the _Optimize Images_ setting is enabled and that _Maximum Image Size_ setting remains in the previously set value.
</details>

<details>
 <summary>Resize settings should function as expected ⤵️</summary>

* Add an image to the device with dimensions above 3000px.
* Remove and install the app.
* Navigate to the Posts/Pages screen.
* Open/create a post/page.
* Add an Image block.
* Tap on the _Choose from device_ option.
* Select the image previously added with large dimensions.
* When the _Optimize Images_ pop-up is displayed, select the option to keep image optimization.
* Wait for the image upload to finish.
* Go to the browser and navigate to the Media screen of the site.
* Select the uploaded image and click on the _Edit_ button.
* Observe that dimensions displayed under the _DIMENSIONS_ section are lower than the original ones. The dimensions shouldn’t exceed the value set in _Max Image Upload Size_ setting.
* Observe that the file size under the _FILE SIZE_ section is smaller than the original file size.</details>

<details>
 <summary>Compression settings should function as expected ⤵️</summary>

- Add an image to the device with a large file size.
- Remove and install the app.
- Navigate to the Posts/Pages screen.
- Open/create a post/page.
- Add an Image block.
- Tap on the _Choose from device_ option.
- Select the image previously added with a large file size.
- When the _Optimize Images_ pop-up is displayed, select the option to keep image optimization.
- Wait for the image upload to finish.
- Go to the browser and navigate to the Media screen of the site.
- Select the uploaded image and click on the _Edit_ button.
- Observe that the file size under the _FILE SIZE_ section is smaller than the original file size.</details>

<details>
<summary>Pop-up should be displayed in specific sources ⤵️</summary>

### Media Screen
- Remove and install the app.
- Navigate to `My Site → Media`.
- Tap on ➕ button and select `Choose from My Device` option.
- Select one or multiple items.
- Tap on the `Add` button.
- Observe that the Optimize Images pop-up is displayed.

### Gutenberg Editor

**Preparation:**
- Remove and install the app.
- Navigate to `My Site → Posts` (or Pages).
- Create a post.

#### Media Blocks - From Device
- Add any of the following blocks:
  - Image block
  - Gallery block
  - Cover block
  - Media & Text block
  - Story block.
- Tap on the `Choose from device` option.
- Select one item (or multiple items when using Gallery block).
- Observe that the Optimize Images pop-up is displayed.

#### Media Blocks - Take Photo

> **Note**
> The pop up doesn't currently display when taking a photo from the Story block. This will be worked on in a future PR.

- Add any of the following blocks:
  - Image block
  - Gallery block
  - Cover block
  - Media & Text block
- Tap on the `Take a Photo` option.
- Take a photo and use it.
- Observe that the Optimize Images pop-up is displayed.

### Featured Image - From Device
- Tap on the `...` button located on the top bar.
- Tap on the `Post Settings` options.
- Navigate to the Featured Image section and tap on `Set Featured Image`.
- Tap on the `Choose from Device` option.
- Select one time.
- Observe that the Optimize Images pop-up is displayed.

### Featured Image - Take Photo
- Tap on the `...` button located on the top bar.
- Tap on the `Post Settings` options.
- Navigate to the Featured Image section and tap on `Set Featured Image`.
- Tap on the `Take a Photo` option.
- Select one time.
- Observe that the Optimize Images pop-up is displayed.</details>

<details>
 <summary>Check events ⤵️</summary>

Verify that the new `APP_SETTINGS_OPTIMIZE_IMAGES_POPUP_TAPPED` event is sent when tapping either of the options in the new popup. `on` will be sent if you leave optimization on and `off` if it's turned off via the popup.
</details>

## Regression Notes
1. Potential unintended areas of impact

As we're changing default settings, there's a risk of unintentionally changing settings users have chosen. We're also changing code around the App Settings screen and media upload flows, so there could be unintended impact there.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

Manually tested each flow we identified and relied on existing tests around media uploads.

3. What automated tests I added (or what prevented me from doing so)

I plan to further analyse existing tests and potentially add new ones in a future PR.

<hr />

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:

- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

SiobhyB committed:
Update strings associated with optimization popup

The strings have been updated to reflect the popup's change in functionality. It will now be informing users about the fact optimization is enabled by default, rather than the fact it used to be off by default.

SiobhyB committed:
Update RELEASE-NOTES

SiobhyB committed:
Set "optimize image" setting to "on" by default

SiobhyB committed:
Update popup functionality to reflect new default

SiobhyB committed:
Show new promo in editor when adding local images

The advertiseImageOptimization() function will now be called when a user uploads a photo from their device.

SiobhyB committed:
Introduce tracks event to monitor clicks

SiobhyB committed:
Display new popup after upload via media screen

SiobhyB committed:
Merge branch 'trunk' into feature/change-default-image-optimization-setting

thomashorta commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/AppPrefs.java:
```diff
@@ -776,7 +776,7 @@ public static void setImageOptimizeQuality(int quality) {
 
     public static int getImageOptimizeQuality() {
-        int quality = getInt(DeletablePrefKey.IMAGE_OPTIMIZE_QUALITY, 0);
+        int quality = getInt(DeletablePrefKey.IMAGE_OPTIMIZE_QUALITY, 1);
```
❓ Unless I'm reading something wrong, having this set to 0 and 1 yields the same result, based on the line immediately below which only uses the `quality` number if it's **greater than** 1, so I believe this change still makes this method return `WPMediaUtils.OPTIMIZE_IMAGE_ENCODER_QUALITY` by default.

Is this change needed?

SiobhyB replied:
Good point! You're right the change isn't required, so I've gone ahead to revert in https://github.com/wordpress-mobile/WordPress-Android/pull/19556/commits/39fbd5362758fffc05cc1d624cd4208d8702730b.

thomashorta commented on WordPress/src/main/java/org/wordpress/android/util/WPMediaUtils.java:
```diff
@@ -142,11 +143,15 @@ public static void advertiseImageOptimization(final Context context,
+                        // null or image optimization already OFF. We should not be here though.
                     } else {
-                        AppPrefs.setImageOptimize(true);
+                        AppPrefs.setImageOptimize(false);
```
🔍 This might be nitpicking but this logic would be clearer if it was just a check leading to an action instead of having the main logic in the `else` clause.

```kotlin
if (AppPrefs.isImageOptimize()) AppPrefs.setImageOptimize(false);
```

SiobhyB replied:
Not a nitpick! I always like the chance to get opinions on improving code quality. :) I'd been following the pattern from the code I was refactoring, but agree that it's not the most straightforward to follow. I like your suggestion and went ahead to implement it in https://github.com/wordpress-mobile/WordPress-Android/pull/19556/commits/ff4fda655c2f961678f294783837a42170deecb2.

SiobhyB committed:
Remove duplicate case in switch statement

Following the last merge from `trunk`, the `STOCK_MEDIA_PICKER_SINGLE_SELECT` case was accidentally duplicated. This PR reverts that mistake.

SiobhyB committed:
Add missing null check

This was accidentally removed when updating from `trunk` in a previous commit.

thomashorta reviewed (commented):
Hi there! Thanks for the changes. I added some minor comments about the code and will now proceed to test the PR behavior just to confirm it before approving.

thomashorta commented on WordPress/src/main/java/org/wordpress/android/ui/media/MediaBrowserActivity.java:
```diff
@@ -499,25 +499,23 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
-                        for (Uri uri : uris) {
-                            getMediaFromDeviceAndTrack(uri, requestCode);
-                        }
+                    if (WPMediaUtils.shouldAdvertiseImageOptimization(this)) {
```
🚫 This check is causing a bug where you can only upload the image when the pop-up appears since it doesn't have an else clause simply calling `handlePickerResult`.

```kotlin
if (WPMediaUtils.shouldAdvertiseImageOptimization(this)) {
    WPMediaUtils.advertiseImageOptimization(this, () -> handlePickerResult(data, resultCode));
} else {
    handlePickerResult(data, resultCode);
}
```

SiobhyB replied:
Ouch, thank you so much for spotting Thomas, can't believe I let that slip through. Addressed in https://github.com/wordpress-mobile/WordPress-Android/pull/19556/commits/29404be05325314ccbeff5623cc423d0106eb6b9. 🤦‍♀️

thomashorta commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -2832,59 +2832,13 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
-                        if (mEditPostSettingsFragment != null) {
-                            mEditPostSettingsFragment.refreshViews();
-                        }
+                    if (WPMediaUtils.shouldAdvertiseImageOptimization(this)) {
```
🚫 Same here. This check is causing a bug where you can only upload the image (in the Post Editor) when the pop-up appears since it doesn't have an else clause simply calling `handlePickerResult`.

```kotlin
if (WPMediaUtils.shouldAdvertiseImageOptimization(this)) {
    WPMediaUtils.advertiseImageOptimization(this, () -> handlePhotoPickerResult(data));
} else {
    handlePhotoPickerResult(data);
}
```

SiobhyB replied:
Addressed in https://github.com/wordpress-mobile/WordPress-Android/pull/19556/commits/29404be05325314ccbeff5623cc423d0106eb6b9, this was a big oversight on my part, thank you for spotting!

thomashorta reviewed (changes requested):
While testing I found a bug that prevents the user from uploading media when the Optimization notice is not shown. I added a couple of relevant comments on the root cause.

SiobhyB committed:
Ensure media is uploaded if popup's not displayed

This commit fixes the bug pinpointed here, in which media wasn't uploading if the popup wasn't displayed: https://github.com/wordpress-mobile/WordPress-Android/pull/19556/files#r1386846666

SiobhyB committed:
Revert unnecessary change to quality default

As noted here, the default yields the same result, whether it's 0 or 1. So, a change is not necessary: https://github.com/wordpress-mobile/WordPress-Android/pull/19556/files#r1386712864

SiobhyB committed:
Refactor popup logic for clarity

Following the review here, the logic surrounding the popup's button has been refactored to make it more clear and readable: https://github.com/wordpress-mobile/WordPress-Android/pull/19556#discussion_r1386733843

SiobhyB commented:
@thomashorta, thank you for the review and for spotting those bugs! I've gone ahead to address your comments 🙇‍♀️

thomashorta reviewed (approved):
I finished reviewing the latest changes and testing all scenarios listed in the PR description and everything looks good to me! Thanks again for these changes @SiobhyB! Nice work on improving the reliability of media uploads! :shipit: 

---

Just a minor note, that we already discussed in a DM: the way we're doing image optimization seems to yield different results depending on the original image being sent, but it looks like it should reduce image size in most cases, though I was able to find one or two pictures that were actually increased in size after passing through optimization (probably due to the compression algorithms that saved those images).

That said, I don't think this would affect the majority of our users who likely upload their own pictures from their phones, which are usually not well compressed in the first place.

SiobhyB committed:
Update wording in popup

The wording has been changed following feedback from the editorial team here: https://github.com/wordpress-mobile/WordPress-iOS/pull/21981#issuecomment-1802410807

SiobhyB commented:
Thank you! I'm going ahead to merge this into a feature branch, `feature/updates-to-optimization-settings`, as we'll likely be merging further changes to the settings as part of this project. 🙇‍♀️
